### ISSUE TITLE
Use `.xignore` file for excludes (Git, etc.)

### ISSUE THREAD

**Issue by sanmai-NL** [editorconfig-checker/editorconfig-checker#50]
I'd say by default.

**Comment by mstruebing** [comment]
`.gitignore` should be respected by default, isn't it? I only check files which are outputted by: `git ls-tree -r --name-only HEAD` if you start the tool inside a git versioned folder.

**Comment by mstruebing** [comment]
I'll close this issue for now. 
If you have additional thoughts or something else, feel free to add your comment or reopen this issue.

**Comment by sanmai-NL** [comment]
@mstruebing 

That doesn't seem to be the case. E.g,, I get as output:

> mypackage/signals/__pycache__/handlers.cpython-39.pyc:
>        Wrong line endings or new final newline
>        Not all lines have the correct end of line character

These files aren't listed in the output from `git ls-tree`.

Can you reopen this issue, please?

**Comment by sanmai-NL** [comment]
See also #99.

**Comment by sanmai-NL** [comment]
@mstruebing The ignores are processed when not providing target paths on the command line, but not when target paths are provided. That does not cut it when you want to scan a specific subdirectory, but still ignore gitignored files under that directory.

**Comment by heygarrett** [comment]
> [...] if you start the tool inside a git versioned folder.

This may not work for [Jujutsu](https://github.com/jj-vcs/jj)-versioned folders (probably depends on whether or not the folder is colocated), but Jujutsu also uses the `.gitignore` file.

**Comment by klaernie** [comment]
So in the end this boils down in my understanding, that we should switch from the plain go file walking to a library like https://github.com/boyter/gocodewalker/ that has built-in support for `.gitignore`.

**Comment by ccoVeille** [comment]
Great idea because I would have said .gitignore is classic, but .ignore too

So I'm happy you found a lib that supports both

**Comment by theoludwig** [comment]
> So in the end this boils down in my understanding, that we should switch from the plain go file walking to a library like [boyter/gocodewalker](https://github.com/boyter/gocodewalker/) that has built-in support for `.gitignore`.

I disagree, we should keep using `git ls-tree -r --name-only HEAD` as it is how `git` really works, if we start using other libraries, we can never be too sure, if it's implemented the "right way", with the same behavior as `git`.
Also, `git`  not only use the `.gitignore` at the root, but also use the `.gitignore` in all the nested directories.

**Comment by theoludwig** [comment]
I had similar concers about other tools, e.g: with ESLint or Prettier, (https://github.com/prettier/prettier/issues/14115).

`editorconfig-checker` is one of the rare tool that implemented it the right way, by using `git` directly instead of reimplementing (or using libraries), the same algorithms used by `git` for ignoring files.

That's one thing I liked about `editorconfig-checker` and I don't want it to change, unless I misunderstood, we should not replace using `git` with https://github.com/boyter/gocodewalker/ or any other libraries or self-implemented algorithm.

**Comment by klaernie** [comment]
We currently have two means of obtaining files in file.go's `GetFiles()`:
https://github.com/editorconfig-checker/editorconfig-checker/blob/main/pkg/files/files.go#L79-L134

Path one handles all files mentioned explicitly on the command line
Path two runs git and adds all files using `git ls-tree`.

My intention was scoped to only the first path - where we currently walk the filetree ourselves and only exclude files by matching against the `config.Exclude` and `config.AllowedContentTypes`.

Replacing the implementation of this first path with a less-manual implementation that respects `.gitignore` files at any depth  means that the runs of `editorconfig-checker` and `editorconfig-checker .` in a directory with a `.gitignore` ignoring files with validations errors will return the same result, same comparing a run of editorconfig-checker on a system lacking git (which then does not have the second path available at all)
